File: COMP26120_DSA/lab3/java/comp26120/model_solutions/bstree.java
package comp26120;

public class bstree implements set<String> {
    int verbose;

    String value;
    bstree left;
    bstree right;

    speller_config config;

    int comparisons = 0;


    public bstree(speller_config config) {
	verbose = config.verbose;
	this.config = config;
    }

	
    public int height() {
	if (tree()) {
	    int lh = this.left.height();
	    int rh = this.right.height();
	    if (lh > rh) {
		return 1 + lh;
	    } else {
		return 1 + rh;
	    }
	}
	return 0;
    }

    
    public int size(){
	// This presumes that if value is not null then (possibly empty) left and right trees exist.
	if(tree()){
	    return 1 + left.size() + right.size();
	}
	return 0;
    }

    public int total_comparisons() {
	if (tree()) {
	    return comparisons + this.left.total_comparisons() + this.right.total_comparisons();
	}
	return 0;
    }

    public void insert (String value) 
    {
	if(tree()){
	    int comp = value.compareTo(this.value);
	    if (comp < 0) {
		this.left.insert(value);
	    } else if (comp > 0) {
		this.right.insert(value);
	    }
	    // if comp == 0 then do nothing as it's a duplicate
	    comparisons++;
	}
	else{
	    left = new bstree(this.config);
	    right = new bstree(this.config);
	    this.value = value;
	}
    }

    public boolean find (String value)
    {
	if(tree()){
	    int comp = value.compareTo(this.value);
	    if (comp == 0) { return true;}
	    else if (comp < 0) {return left.find(value);}
	    else return right.find(value);
	}
	// if tree is NULL then it contains no values
	return false;
    }

    private boolean tree() {
	return (value != null);
    }

    // You can update this if you want
    public void print_set_recursive(int depth)
    {
	if(tree()){
	    for(int i=0;i<depth;i++){ System.out.print(" "); }
	    System.out.format("%s\n",value);
	    left.print_set_recursive(depth+1);
	    right.print_set_recursive(depth+1);
	}
    } 

    // You can update this if you want
    public void print_set ()
    {
	System.out.print("Tree:\n");
	print_set_recursive(0);
    }

    public void print_stats ()
    {
	System.out.println("Height: " + this.height());
	System.out.println("Comparisons per insert: " + this.total_comparisons()/this.size());
    }
}

File: COMP26120_DSA/lab3/python/model_solutions/bstree.py
class bstree:
    def __init__(self):
        pass
        
    def size(self):
        if (self.tree()):
            return 1 + self.left.size() + self.right.size()
        return 0
        
    def height(self):
        if (self.tree()):
            lh = self.left.height()
            rh = self.right.height()
            if (lh > rh):
                return 1 + lh
            else:
                return 1 + rh
        return 0
        
    def total_comparisons(self):
        if (self.tree()):
            return self.comparisons + self.left.total_comparisons() + self.right.total_comparisons()
        return 0
        
    def tree(self):
        return hasattr(self, 'value')
        
    def insert(self, value):
        if (self.tree()):
            if(value < self.value):
                self.left.insert(value)
            elif (value > self.value):
                self.right.insert(value)
            self.comparisons = self.comparisons + 1
        else:
            self.left = bstree()
            self.right = bstree()
            self.value = value
            self.comparisons = 0
        
    def find(self, value):
        if self.tree():
            if (value == self.value):
                return True
            elif (value < self.value):
                return self.left.find(value)
            else:
                return self.right.find(value)
        return False
        
    # You can update this if you want
    def print_set_recursive(self, depth):
        if (self.tree()):
            for i in range(depth):
                print(" ", end='')
            print("%s" % self.value)
            self.left.print_set_recursive(depth + 1)
            self.right.print_set_recursive(depth + 1)
            
    # You can update this if you want
    def print_set(self):
        print("Tree:\n")
        self.print_set_recursive(0)
        
    def print_stats(self):
        print("Height: %d" % self.height())
        print("Comparisions per insert: %s" % str(self.total_comparisons()/self.size()))
             
            
